### PR TITLE
chore: Refactor credential rotator into a generic function

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/handler.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    #   ]
+    #   "destinations": {
+    #       "aws-sdk-ios-cci": {
+    #           "type": "cci_env_variable",
+    #           "description": "Circle CI environment variable for AWS SDK iOS repo",
+    #           "github_path": "aws-amplify/aws-sdk-ios",
+    #           "secret_arn_env_variable": "CIRCLE_CI_IOS_SDK_API_TOKEN"            
+    #        }
+    #   }
+    # }
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/handler.py:81]: JSON syntax. Please review `mapping_to_destination` -- I think we decided that was just a straight-up map, not an array of objects?:

```suggestion
    """
    Invoked with the following event structure:
    ```
    {
      "sources": [
        {
          "type": "aws_session_cred",
          "description": "Temporary AWS Credentials to upload the release artifacts to S3 and invalidate Cloudfront",
          "configuration": {
            "user_env_variable": "IAM_USERNAME",
            "iam_role_env_variable": "IAM_ROLE"
          },
          "destination": {
            "specifier": "aws-sdk-ios-cci",
            "mapping_to_destination": [
              {
                "destination_key_name": "XCF_ACCESS_KEY_ID",
                "result_value_key": "access_key_id"
              },
              {
                "destination_key_name": "XCF_SECRET_ACCESS_KEY",
                "result_value_key": "secret_access_key"
              },
              {
                "destination_key_name": "XCF_SESSION_TOKEN",
                "result_value_key": "session_token"
              }
            ]
          }
        },
        {
          "type": "secrets_manager",
          "description": "",
          "configuration": {
            "secret_key_env_variable": "GITHUB_CREDENTIALS_SECRET"
          },
          "destination": {
            "specifier": "aws-sdk-ios-cci",
            "mapping_to_destination": [
              {
                "destination_key_name": "GITHUB_SPM_TOKEN",
                "result_value_key": "GITHUB_SPM_TOKEN"
              },
              {
                "destination_key_name": "GITHUB_SPM_USER",
                "result_value_key": "GITHUB_SPM_USER"
              }
            ]
          }
        },
        {
          "type": "lambda_env_variables",
          "description": "",
          "configuration": {
            "lambda_env_var_key": "SPM_S3_BUCKET_NAME"
          },
          "destination": {
            "specifier": "aws-sdk-ios-cci",
            "mapping_to_destination": [
              {
                "destination_key_name": "XCF_S3_BUCKET_NAME"
              }
            ]
          }
        }
      ],
      "destinations": {
        "aws-sdk-ios-cci": {
          "type": "cci_env_variable",
          "description": "Circle CI environment variable for AWS SDK iOS repo",
          "github_path": "aws-amplify/aws-sdk-ios",
          "secret_arn_env_variable": "CIRCLE_CI_IOS_SDK_API_TOKEN"
        }
      }
    }
    ```
    """

```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/handler.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        configuration = source["configuration"]
+        
+        if source_type == SourceType.AWS_SESSION_CREDENTIALS:
+            mapped_result = aws_credential_rotator.generate_session_credentials(
+                configuration,
+                destination_mapping=destination_mapping
+            )
+            
+        elif source_type ==  SourceType.SECRETS_MANAGER:
+            mapped_result = {}
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/handler.py:100]: I assume this and the LAMBDA_ENVIRONMENT_VARIABLE is coming in an upcoming PR?

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/models/destination_type.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from enum import Enum
+
+class DestinationType(Enum):
+
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/models/destination_type.py:4]:
```suggestion
    """A destination that rotated values are pushed to
    """
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/models/source_type.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from enum import Enum
+
+class SourceType(Enum):
+
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/models/source_type.py:4]:
```suggestion
    """The source of a value to push to a destination
    """
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/source_data_generator/aws_credential_rotator.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
+import boto3
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/source_data_generator/aws_credential_rotator.py:1]: 1. This is now not really an `aws_credential_rotator` so much as an `aws_session_credential_source` or similar, right?
2. Doc comment:
```suggestion
"""Creates static credentials, uses them to adopt the specified role and get temporary session credentials, and deletes the static credentials.

import boto3
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/source_data_generator/aws_credential_rotator.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+# test is invoked with session credentials created with access keys that are rotated immediately
+# thereafter.
+SESSION_DURATION_SECONDS = 14400  # 4 hours
+
+# Provided by Lambda
+REGION = os.environ.get("AWS_REGION", DEFAULT_REGION)
+
+def generate_session_credentials(
+    configuration: map, 
+    destination_mapping: list,
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/source_data_generator/aws_credential_rotator.py:22]: Let's not have the mapping as part of this logic; the logic to generate & return temporary session credentials doesn't need to know that they'll be mapped to different values elsewhere

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/test/test_aws_credential_rotator.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+import unittest
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/test/test_aws_credential_rotator.py:1]: !!! Tests!

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/destination/circleci.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+
+import os
+import boto3
+import requests
+from src.utils.retry import retry
+
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/destination/circleci.py:30]: Ran the linting

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/destination/circleci.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
+"""Utilities to update CircleCI environment variables
+
+This module propagates values to CircleCI environment variables. In the CircleCI V2 API
+
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/destination/circleci.py:4]: remove fragment

```suggestion
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/handler.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+def handler(event, context, *, iam=None, sts=None, secretsmanager=None):
+    """
+    Invoked with the following event structure:
+    ```
+    {
+      "sources": [
+        {
+          "type": "aws_session_cred",
+          "description": "Temporary AWS Credentials to upload the release
+                          artifacts to S3 and invalidate Cloudfront",
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/handler.py:18]: Multi-line JSON string not allowed. I know it's a comment, but it'd be nice to make this copy/pasteable

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/source_data_generator/aws_session_credential_source.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+        iam_client = iam or boto3.client("iam", region_name=REGION)
+        user_credentials = create_user_credentials(iam_username, iam=iam_client)
+        wait_for_user_credentials()
+        session_credentials = get_session_credentials(user_credentials, role_arn=iam_role, sts=sts)
+
+    finally:
+        if user_credentials:
+            iam_client.delete_access_key(UserName=iam_username, AccessKeyId=user_credentials[0])
+    return session_credentials
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/src/source_data_generator/aws_session_credential_source.py:49]: nit, add vertical whitespace
```suggestion

    return session_credentials
```

**Review by palpatim — APPROVED** [review]
Minor formatting comments, otherwise LGTM!

**Pull request merged**